clusterloader2/pkg/framework/client/objects.go

```diff
@@ -110,20 +113,19 @@ func CreateObject(dynamicClient dynamic.Interface, namespace string, name string
 
 // UpdateObject updates with given name, group, version and kind based on given object description.
 func UpdateObject(dynamicClient dynamic.Interface, namespace string, name string, obj *unstructured.Unstructured) error {
-	// TODO(krzysied): This method should be more generic and robust.
 	gvk := obj.GroupVersionKind()
 	gvr, _ := meta.UnsafeGuessKindToResource(gvk)
 	currentObj, err := GetObject(dynamicClient, gvk, namespace, name)
 	if err != nil {
 		return err
 	}
-	spec, exists := obj.UnstructuredContent()["spec"]
-	if !exists {
-		return fmt.Errorf("Object without a spec")
+	obj.SetName(name)
+	patch, err := createPatch(currentObj, obj)
+	if err != nil {
+		return fmt.Errorf("creating patch diff error: %v", err)
 	}
-	unstructured.SetNestedField(currentObj.UnstructuredContent(), spec, "spec")
 	updateFunc := func() error {
-		_, err := dynamicClient.Resource(gvr).Namespace(namespace).Update(currentObj)
+		_, err := dynamicClient.Resource(gvr).Namespace(namespace).Patch(obj.GetName(), types.StrategicMergePatchType, patch)
```

Comment:
Please move the above code block where you're computing the patch into this updateFunc. Otherwise I think the logic would be wrong, because you need to recalculate the patch each time you failed to update the object due to a conflict. RIght now you're re-using the same patch again and again.

Reply:
Get object function and createPatch moved into retryable function.
I agree that's better approach, although previous one also worked due to the assumption that there are no external things operating on the same object.

Reply:
That's not a very good assumption to make in practice :)
Because there are so many interactions in the system and components that we aren't even aware of may potentially change the objects.